api/utils/keys/yubikey.go

```diff
@@ -0,0 +1,407 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer yk.Close()
+
+	opts := piv.Key{
+		Algorithm:   piv.AlgorithmEC256,
+		PINPolicy:   piv.PINPolicyNever,
+		TouchPolicy: touchPolicy,
+	}
+	pub, err := yk.GenerateKey(piv.DefaultManagementKey, slot, opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create a self signed certificate and store it in the PIV slot so that other
+	// Teleport Clients know to reuse the stored key instead of genearting a new one.
+	priv, err := yk.PrivateKey(slot, pub, piv.KeyAuth{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cert, err := selfSignedTeleportClientCertificate(priv, pub)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Store an unsigned certificate to mark this slot as used by tsh.
+	if err = yk.SetCertificate(piv.DefaultManagementKey, slot, cert); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return newYubikeyPrivateKey(y, slot, pub)
+}
+
+// getPrivateKey gets an existing private key from the given PIV slot.
+func (y *yubikey) getPrivateKey(slot piv.Slot) (*YubikeyPrivateKey, error) {
+	yk, err := y.open()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer yk.Close()
+
+	// Check the slot's certificate to see if it contains a self signed Teleport Client cert.
+	cert, err := yk.Certificate(slot)
+	if err != nil || cert == nil {
+		return nil, trace.NotFound("Yubikey certificate slot is empty, expected a Teleport Client cert")
+	} else if len(cert.Subject.Organization) == 0 || cert.Subject.Organization[0] != certOrgName {
+		return nil, trace.NotFound("Yubikey certificate slot contained unknown certificate:\n%+v", cert)
+	}
+
+	// Attest the key to make sure it hasn't been imported.
+	slotCert, err := yk.Attest(slot)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	attestationCert, err := yk.AttestationCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if _, err = piv.Verify(attestationCert, slotCert); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Verify that the slot's certs have the same public key, otherwise the key
+	// may have been generated by a non-teleport client.
+	if !cert.PublicKey.(interface{ Equal(x crypto.PublicKey) bool }).Equal(slotCert.PublicKey) {
+		return nil, trace.NotFound("Yubikey slot contains mismatched certificates and must be regenerated")
+	}
+
+	return newYubikeyPrivateKey(y, slot, slotCert.PublicKey)
+}
+
+// open a connection to yubikey PIV module. The returned connection should be closed once
+// it's been used. The yubikey PIV module itself takes some additional time to handle closed
+// connections, so we use a retry loop to give the PIV module time to close prior connections.
+func (y *yubikey) open() (yk *piv.YubiKey, err error) {
+	isRetryError := func(err error) bool {
+		retryError := "connecting to smart card: the smart card cannot be accessed because of other connections outstanding"
+		return strings.Contains(err.Error(), retryError)
+	}
+
+	var maxRetries int = 100
```

Comment:
This can block for quite a long time and there's no way to cancel it either - is it worth it to pass along some context or is this only ever going to be used by tsh/Connect so it doesn't matter?

Even in that case, blocking for up to 10 seconds before closing the program seems a bit of a UI blunder.

Reply:
> This can block for quite a long time and there's no way to cancel it either - is it worth it to pass along some context or is this only ever going to be used by tsh/Connect so it doesn't matter?

This would require adding `ctx` to `crypto.Signer.Sign`, which is not possible while maintaining functionality as a `crypto.Signer`. I think the only solution we can introduce is to reduce the retry limit to (ideally) under 100 milliseconds (which we can't do until the below issue is address)

> Even in that case, blocking for up to 10 seconds before closing the program seems a bit of a UI blunder.

Retries primarily occur when a PIV connection is opened during or immediately after a previous connection. A short backoff is sufficient for `tsh` which execute sequentially, but Teleport Connect requires a longer backoff since it opens the PIV private key countless times on startup.

Since PIV devices only allow a single connection at a time, the primary way we can improve this is by reusing connections. Since any Teleport process can open a PIV connection at any time to access it's associated private key, we would need to maintain a global connection pool with cached clients that are closed after a delay. 

I'm not sure of the best way to implement this and am concerned that it will block the PR, so I was planning on making this improvement in a follow up PR (after https://github.com/gravitational/teleport/pull/15336 which introduces the remaining core functionality).

Reply:
And we'd need some retry logic for the first opening of the key anyway, unless we're fine with just erroring out and exiting 😢

Reply:
Yes :( We could potentially open the yubikey ahead of time under a `ctx`, but managing the connection and closing it when it's not being used would become a struggle. For example, while the PIV connection is held open, the yubikey cannot accept MFA requests.